client/client.go

```diff
@@ -1,29 +1,401 @@
+func (client *CqlClient) ConnectAndInit(version primitive.ProtocolVersion) (*CqlClientConnection, error) {
+	if connection, err := client.Connect(); err != nil {
 		return nil, err
+	} else {
+		return connection, connection.InitiateHandshake(version)
+	}
+}
+
+// CqlClientConnection encapsulates a TCP client connection to a remote Cassandra-compatible backend.
+// CqlClientConnection instances should be created by calling CqlClient.Connect or CqlClient.ConnectAndInit.
+type CqlClientConnection struct {
+	conn            net.Conn
+	codec           frame.Codec
+	readTimeout     time.Duration
+	credentials     *AuthCredentials
+	inFlightHandler *inFlightRequestsHandler
+	outgoing        chan *frame.Frame
+	events          chan *frame.Frame
+	waitGroup       *sync.WaitGroup
+	closed          int32
+}
+
+func newCqlClientConnection(
+	conn net.Conn,
+	credentials *AuthCredentials,
+	codec frame.Codec,
+	maxInFlight int,
+	maxPending int,
+	readTimeout time.Duration,
+) (*CqlClientConnection, error) {
+	if maxInFlight < 1 {
+		return nil, fmt.Errorf("max in-flight: expecting positive, got: %v", maxInFlight)
+	} else if maxInFlight > math.MaxInt16 {
+		return nil, fmt.Errorf("max in-flight: expecting <= %v, got: %v", math.MaxInt16, maxInFlight)
+	}
+	if maxPending < 1 {
+		return nil, fmt.Errorf("max pending: expecting positive, got: %v", maxInFlight)
+	}
+	if codec == nil {
+		codec = frame.NewCodec()
+	}
+	connection := &CqlClientConnection{
+		conn:        conn,
+		codec:       codec,
+		readTimeout: readTimeout,
+		credentials: credentials,
+		outgoing:    make(chan *frame.Frame, maxInFlight),
+		events:      make(chan *frame.Frame, maxInFlight),
+		waitGroup:   &sync.WaitGroup{},
+	}
+	connection.inFlightHandler = newInFlightRequestsHandler(connection.String(), maxInFlight, maxPending, readTimeout)
+	connection.incomingLoop()
+	connection.outgoingLoop()
+	return connection, nil
+}
+
+func (c *CqlClientConnection) String() string {
+	return fmt.Sprintf("CQL client conn [L:%v <-> R:%v]", c.conn.LocalAddr(), c.conn.RemoteAddr())
+}
+
+func (c *CqlClientConnection) incomingLoop() {
+	log.Debug().Msgf("%v: listening for incoming frames...", c)
+	c.waitGroup.Add(1)
+	go func() {
+		defer c.waitGroup.Done()
+		for !c.isClosed() {
+			if incoming, err := c.codec.DecodeFrame(c.conn); err != nil {
+				if !c.isClosed() {
+					if errors.Is(err, io.EOF) {
+						log.Info().Msgf("%v: connection reset by peer, closing", c)
+					} else {
+						log.Error().Err(err).Msgf("%v: error reading, closing connection", c)
+					}
+					c.abort()
+				}
+				break
+			} else {
+				log.Debug().Msgf("%v: received incoming frame: %v", c, incoming)
+				if incoming.Header.OpCode == primitive.OpCodeEvent {
+					select {
+					case c.events <- incoming:
+						log.Debug().Msgf("%v: incoming event frame successfully delivered: %v", c, incoming)
+					default:
+						log.Error().Msgf("%v: incoming event frame delivery failed: %v", c, incoming)
```

Comment:
could use a better error message, something like "events queue is full, discarding event message"